File: packages/sdk-platforms/swift/ZKsyncSSO/Sources/ZKsyncSSO/AccountClient.swift
import Foundation
@preconcurrency import ZKsyncSSOFFI

public struct AccountClient: Sendable {
    
    public let authenticatorAsync: any PasskeyAuthenticatorAsync & Sendable
    
    public let account: Account
    
    public init(
        account: Account,
        authenticatorAsync: any PasskeyAuthenticatorAsync & Sendable
    ) {
        self.account = account
        self.authenticatorAsync = authenticatorAsync
    }
    
    public func getAccountBalance() async throws -> String {
        let accountBalance = try await ZKsyncSSOFFI.getBalance(
            address: account.address,
            config: Config.default.inner
        )
        return accountBalance.balance
    }
    
    public func fundAccount() async throws {
        try await ZKsyncSSOFFI.fundAccount(
            address: account.address,
            config: Config.default.inner
        )
    }
    
    public func sendTransaction(
        to: String,
        amount: String
    ) async throws {
        let tx = Transaction(
            from: account.address,
            to: to,
            value: amount,
            input: nil
        )
        let result = try await ZKsyncSSOFFI.sendTransactionAsyncSigner(
            transaction: tx,
            authenticator: authenticatorAsync,
            config: Config.default.inner
        )
        print(result)
    }
    
    public func prepareTransaction(
        transaction: TransactionRequest
    ) async throws -> PreparedTransaction {
        let from = account.address
        let tx = try await ZKsyncSSOFFI.prepareSendTransaction(
            transaction: transaction.inner,
            from: from,
            config: Config.default.inner
        )
        return tx.wrappedValue
    }
}

File: packages/sdk-platforms/react-native/react-native-zksync-sso/src/passkey/authenticate/account_client.ts
// @ts-ignore
import {
    type Config,
    type Account,
    type PreparedTransaction,
    type Transaction,
    type SendTransactionResult,
    sendTransactionAsyncSigner,
    prepareSendTransaction,
} from 'react-native-zksync-sso';
import { Authenticator } from './authenticator';
export { type PreparedTransaction };

/**
 * Helper class for account operations like transaction preparation and sending
 */
export class AccountClient {
    private account: Account;
    private rpId: string;
    private config: Config;

    constructor(account: Account, rpId: string, config: Config) {
        this.account = account;
        this.rpId = rpId;
        this.config = config;
    }

    /**
     * Prepares a transaction for sending
     * @param transaction The transaction to prepare
     * @returns Prepared transaction with fee information
     */
    async prepareTransaction(to: string, value: string): Promise<PreparedTransaction> {
        const from = this.account.address;
        const transaction: Transaction = {
            to,
            value,
            from,
        };
        console.log("transaction: ", transaction);
        const preparedTransaction = await prepareSendTransaction(
            transaction,
            from,
            this.config
        );
        console.log("preparedTransaction: ", preparedTransaction);
        return preparedTransaction;
    }

    /**
     * Sends a transaction
     * @param to Recipient address
     * @param amount Amount in wei as string
     * @returns Transaction hash
     */
    async sendTransaction(to: string, value: string): Promise<SendTransactionResult> {
        const authenticator = new Authenticator(this.rpId);
        const prepared = await this.prepareTransaction(to, value);
        const result = await sendTransactionAsyncSigner(
            prepared,
            authenticator,
            this.config
        );
        console.log("result: ", result);
        return result;
    }
}